config.md

```diff
@@ -384,31 +384,80 @@ For POSIX platforms, the configuration structure supports `hooks` for configurin
 
 ### <a name="configHooksPrestart" />Prestart
 
-The pre-start hooks MUST be called after the [`start`](runtime.md#start) operation is called but [before the user-specified program command is executed](runtime.md#lifecycle).
+The `prestart` hooks MUST be called after the [`start`](runtime.md#start) operation is called but [before the user-specified program command is executed](runtime.md#lifecycle).
 On Linux, for example, they are called after the container namespaces are created, so they provide an opportunity to customize the container (e.g. the network namespace could be specified in this hook).
 
+Note: `prestart` hooks were deprecated in favor of `create-runtime`, `create-container` and `start-container` hooks, which allow more granular hook control during the create and start phase.
+`prestart` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
+
+### <a name="configHooksCreateRuntime" />Create Runtime Hooks
+
+The `create-runtime` hooks MUST be called as part of the [`create`](runtime.md#create) operation after the runtime environment has been created (according to the configuration in config.json) but before the `pivot_root` or any equivalent operation has been executed.
+The `create-runtime` hooks MUST be executed in the [runtime namespace](glossary.md#runtime-namespace).
+
+On Linux, for example, they are called after the container namespaces are created, so they provide an opportunity to customize the container (e.g. the network namespace could be specified in this hook).
+
+Note: `runc` originally implemented `prestart` hooks contrary to the spec, namely as part of the `create` operation (instead of during the `start` operation). This incorrect implementation actually corresponds to `create-runtime` hooks. For runtimes that implement the deprecated `prestart` hooks as `create-runtime` hooks, `create-runtime` hooks MUST be called after the `prestart` hooks.
+
+### <a name="configHooksCreateContainer" />Create Container Hooks
+
+The `create-container` hooks MUST be called as part of the [`create`](runtime.md#create) operation after the runtime environment has been created (according to the configuration in config.json) but before the `pivot_root` or any equivalent operation has been executed.
+The `create-container` hooks MUST be executed in the [container namespace](glossary.md#container-namespace).
+The `create-container` hooks MUST be called after the `create-runtime` hooks.
+
+For example, on Linux this would happen before the `pivot_root` operation is executed but after the mount namespace was created and setup.
+
+### <a name="configHooksStartContainer" />Start Container Hooks
+
+The `start-container` hooks MUST be called [before the user-specified process is executed](runtime.md#lifecycle) as part of the [`start`](runtime.md#start) operation.
```

Comment:
From the description, this sounds like a pre exec hook. I understand the use case but wonder if we can name it something else.

Reply:
IIUC, this is indeed what the `prestart` hooks were originally specced to be (but which never got implemented in this way). Due to the deprecation, `prestart` cannot be reused as a name, while the suggested names follow a `<operation>-<namespace>` pattern.

The easiest adaptation might be to re-add the "pre"-prefix, i.e. `pre<operation>-<namespace>`, yielding `precreate-runtime`, `precreate-container` & `prestart-container`, but I'm not sure that's better (since the hook is executed _during_  rather than _before_ `<operation>`). Of course, there's no fundamental reason why this hook could not be named completely differently, just that the bikeshedding has gone on for a long time already. ;-)